### PR TITLE
ssh/tunnel: add ssh tunnel that implement net.Conn

### PR DESCRIPTION
**Pull request by sitano**
SSH tunnel implementing net.Conn provides a means for having HTTP+SSH clients.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,7 +4,8 @@ Package sshtools provides a wrapper around SSH client with the following feature
 
 * Context aware (dial and execution),
 * Keepalive enabled,
-* Can copy files using SCP.
+* Can copy files using SCP,
+* Provide SSH tunneling for net.Conn (HTTP+SSH).
```
Comment (mmatczuk) [inline README.md:8]: `Provide` is not a feature

File: communicator.go [hunk 2]
```diff
@@ -31,6 +31,7 @@ type Logger interface {
 type Communicator struct {
 	host   string
 	config Config
+
```
Comment (mmatczuk) [inline communicator.go:34]: Why?

File: communicator.go [hunk 3]
```diff
@@ -79,7 +80,7 @@ func (c *Communicator) Connect(ctx context.Context) (err error) {
 	if c.config.KeepaliveEnabled() {
 		c.logger.Println("Starting ssh KeepAlives", "host", c.host)
 		c.keepaliveDone = make(chan struct{})
-		go StartKeepalive(client, c.config.ServerAliveInterval, c.config.ServerAliveCountMax, c.keepaliveDone)
+		go KeepAlive(client, c.config.ServerAliveInterval, c.config.ServerAliveCountMax, c.keepaliveDone)
```
Comment (mmatczuk) [inline communicator.go:83]: Why?

File: communicator.go [hunk 4]
```diff
@@ -97,7 +98,7 @@ func (c *Communicator) reset() {
 	c.keepaliveDone = nil
 
 	if c.client != nil {
-		c.client.Close()
+		_ = c.client.Close()
```
Comment (mmatczuk) [inline communicator.go:101]: Why?

File: dial.go [hunk 5]
```diff
@@ -27,34 +39,26 @@ func (d contextDialer) DialContext(ctx context.Context, network, addr string, co
 		return nil, err
 	}
 
-	type dialRes struct {
-		client *ssh.Client
-		err    error
-	}
-	dialc := make(chan dialRes, 1)
-
+	var client *ssh.Client
```
Comment (mmatczuk) [inline dial.go:42]: I fail to understand purpose of this change. This function in design and naming follows http package no need to butcher it.

File: keepalive.go [hunk 6]
```diff
@@ -8,24 +8,27 @@ import (
...
 		select {
 		case <-t.C:
-			if err := serverAliveCheck(client); err != nil {
-				n++
-				if n >= countMax {
-					client.Close()
-					return
-				}
-			} else {
+			if err := serverAliveCheck(client); err == nil {
```
Comment (mmatczuk) [inline keepalive.go:22]: This improves readability, please extract to separate commit.

File: keepalive.go [hunk 7]
```diff
@@ -8,24 +8,27 @@ import (
 	"golang.org/x/crypto/ssh"
 )
 
-// StartKeepalive starts sending server keepalive messages until done channel
-// is closed.
-func StartKeepalive(client *ssh.Client, interval time.Duration, countMax int, done <-chan struct{}) {
+// KeepAlive keeps an ssh channel alive sending keepalive pings
+// every interval seconds. The channel outlives up to maxErrors consecutive failures.
+// Done channel is a control channel that shutdowns a sidecar goroutine.
+func KeepAlive(client *ssh.Client, interval time.Duration, maxErrors int, done <-chan struct{}) {
```
Comment (mmatczuk) [inline keepalive.go:14]: countMax is not some random string it relates to `ServerAliveCountMax` ssh option.

File: proxydialer.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/pkg/errors"
+)
+
+// ProxyDialer is a dialler that allows for proxying connections over SSH.
+type ProxyDialer struct {
```
Comment (mmatczuk) [inline proxydialer.go:14]: I suggest we move that to separate package `httptun` maybe

File: proxydialer.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+// NewProxyDialer creates an instance of a ProxyDialer that is capable of
+// setting up an ssh tunnels.
+func NewProxyDialer(config Config, dial DialContextFunc) *ProxyDialer {
+	return &ProxyDialer{config: config, dial: dial}
+}
+
+// DialContext dials to addr (HOST:PORT) and establishes an SSH connection
+// to HOST and then proxies the connection to localhost:PORT.
+func (pd *ProxyDialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
```
Comment (mmatczuk) [inline proxydialer.go:30]: Why not add the docs you added to the other DialContext func?

File: proxydialer.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+	host, port, err := net.SplitHostPort(addr)
+	defer func() {
+		if pd.OnDial != nil {
+			pd.OnDial(host, err)
+		}
+	}()
+
+	tun.client, err = pd.dial(ctx, network,
+		net.JoinHostPort(host, fmt.Sprint(pd.config.Port)), &pd.config.ClientConfig)
```
Comment (mmatczuk) [inline proxydialer.go:41]: Maybe extract to var and use single line?

File: keepalive.go (hunk 7 above)
Comment (sitano) [reply keepalive.go:14]: You right. What do you think if we rename `ServerAliveCountMax` to `ServerAliveMaxErrors`? It may break compatibility, but just looking onto the name itself it seems to be it does not deliver its meaning

File: communicator.go (hunk 4 above)
Comment (sitano) [reply communicator.go:101]: I have linters that warn all the time that we ignore returned error

File: doc.go [hunk 11]
```diff
@@ -6,4 +6,5 @@ package sshtools
 //
 // * Context aware (dial and execution),
 // * Keepalive enabled,
-// * Can copy files using SCP.
+// * Can copy files using SCP,
+// * Provide SSH tunneling for net.Conn (HTTP+SSH).
```
Comment (mmatczuk) [inline doc.go:10]: Same as earlier.

File: communicator.go (hunk 3 above)
Comment (sitano) [reply communicator.go:83]: I thought that KeepAlive is self-sufficient to describe what type of action it implies.
KeepAlive is more like a feature-adder, so I wanted underscore its semantic meaning rather than functional.

File: keepalive.go (hunk 7 above)
Comment (mmatczuk) [reply keepalive.go:14]: Sure send PR to OpenSSH if they merge it I'm fine with that

File: tunnel.go [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+package sshtools
+
+import (
+	"net"
+
+	"golang.org/x/crypto/ssh"
+)
+
+// Tunnel holds an ssh tunnel to some resource on remote machine
+type Tunnel struct {
```
Comment (mmatczuk) [inline tunnel.go:12]: I'd suggest renaming that to Conn when moved to separate package.

File: tunnel.go [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+// Client returns ssh client instance.
+func (c Tunnel) Client() *ssh.Client {
+	return c.client
+}
+
+// Close closes the connection and frees the associated resources.
+func (c Tunnel) Close() error {
+	defer c.free()
+
+	if c.done != nil {
```
Comment (mmatczuk) [inline tunnel.go:32]: L32-EOF should be done in free.

File: tunnel.go [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+	client *ssh.Client
+
+	done chan struct{}
+	free func()
+}
+
+var _ net.Conn = (*Tunnel)(nil)
+
+// Client returns ssh client instance.
+func (c Tunnel) Client() *ssh.Client {
```
Comment (mmatczuk) [inline tunnel.go:24]: Why this is needed / useful?

**Review by mmatczuk — CHANGES_REQUESTED** [review]
I'd suggest to
* Extract relevant refactorings to separate commits
* Extract the tunnel part to separate package

some comments included inline

File: communicator.go (hunk 3 above)
Comment (mmatczuk) [reply communicator.go:83]: We can rename that in a separate commit
